src/RouterView.ts

```diff
@@ -34,80 +36,66 @@ export const RouterViewImpl = defineComponent({
-      if (componentProps === true) return route.value.params
+    provide(viewDepthKey, depth + 1)
+    provide(matchedRouteKey, matchedRouteRef)
 
-      return typeof componentProps === 'object'
-        ? componentProps
-        : componentProps(route.value)
-    })
+    const viewRef = ref<ComponentPublicInstance>()
 
-    provide(matchedRouteKey, matchedRoute)
+    return () => {
+      const matchedRoute = matchedRouteRef.value
+      if (!matchedRoute) {
+        return null
+      }
 
-    const viewRef = ref<ComponentPublicInstance>()
+      const ViewComponent = matchedRoute.components[props.name]
+      if (!ViewComponent) {
+        return null
+      }
 
-    function onVnodeMounted() {
-      // if we mount, there is a matched record
-      matchedRoute.value!.instances[props.name] = viewRef.value
-      // TODO: trigger beforeRouteEnter hooks
-      // TODO: watch name to update the instance record
-    }
+      // props from route configration
+      const routePropsOption = matchedRoute.props[props.name]
+      const routeProps = routePropsOption
+        ? routePropsOption === true
+          ? route.params
+          : typeof routePropsOption === 'function'
+          ? routePropsOption(route)
+          : routePropsOption
+        : null
 
-    return () => {
       // we nee the value at the time we render because when we unmount, we
       // navigated to a different location so the value is different
-      const currentMatched = matchedRoute.value
       const currentName = props.name
-      function onVnodeUnmounted() {
-        if (currentMatched) {
-          // remove the instance reference to prevent leak
-          currentMatched.instances[currentName] = null
-        }
+      const onVnodeMounted = () => {
+        matchedRoute.instances[currentName] = viewRef.value
+        // TODO: trigger beforeRouteEnter hooks
+      }
+      const onVnodeUnmounted = () => {
+        // remove the instance reference to prevent leak
+        matchedRoute.instances[currentName] = null
       }
 
-      let Component = ViewComponent.value
-      const componentProps: Parameters<typeof h>[1] = assign(
-        {},
-        // only compute props if there is a matched record
-        Component && propsData.value,
-        attrs,
-        {
+      const component = h(
+        ViewComponent,
+        assign({}, routeProps, attrs, {
           onVnodeMounted,
           onVnodeUnmounted,
           ref: viewRef,
-        }
+        })
       )
 
-      // NOTE: we could also not render if there is no route match
-      const children =
-        slots.default && slots.default({ Component, props: componentProps })
-
-      return children
-        ? children
-        : Component
-        ? h(Component, componentProps)
-        : null
+      return (
+        // pass the vnode to the slot as a prop.
+        // h and <component :is="..."> both accept vnodes
+        (slots.default && slots.default({ Component: component })) || component
```

Comment:
I think here we could just use a ternary:

```suggestion
        slots.default ? slots.default({ Component: component }) : component
```

Reply:
Aah, I didn't see the parentheses correctly. But I think it would make more sense to use the slot if there is one without falling back to `component`. This way the user could hide it if they wanted to

Reply:
Yeah I just followed the original logic. We should probably use a ternary indeed.